File: src/baekjoon/dp/b2931/b2931.java
package baekjoon.dp.b2931;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class b2931 {
    static char[][] map;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int R = Integer.parseInt(st.nextToken());
        int C = Integer.parseInt(st.nextToken());
        map = new char[R][C];

        int startR = -1, startC = -1;
        for (int i = 0; i< R; i++) {
            String line = br.readLine();
            for (int j = 0; j < C; j++) {
                map[i][j] = line.charAt(j);
                if (map[i][j] == 'M') {
                    startR = i;
                    startC = j;
                }
            }
        }
        // 1. 첫 흐름 찾기
        int r = -1, c = -1, dir = -1;
        for (int i = 0; i < 4; i++) {
            int nr = startR + dr[i];
            int nc = startC + dc[i];
            if (nr >= 0 && nr < R && nc >= 0 && nc < C && map[nr][nc] != '.' && map[nr][nc] != 'Z') {
                r = nr;
                c = nc;
                dir = i;
                break;
            }
        }

        // 2. 시뮬레이션
        while (map[r][c] != '.') {
            char block = map[r][c];

            // 현재 진행 방향(dir)과 블록 종류에 따라 다음 진행 방향(nextDir)을 결정
            if (block == '1') {
                dir = (dir == 0) ? 3 : 1;
            } else if (block == '2') {
                dir = (dir == 1) ? 3 : 0;
            } else if (block == '3') {
                dir = (dir == 1) ? 2 : 0;
            } else if (block == '4') {
                dir = (dir == 0) ? 2 : 1;
            }
            // block이 '|', '-', '+' 일때는 방향이 바뀌지 않으므로 dir을 그대로 사용

            // 계산된 다음 방향으로 이동
            r += dr[dir];
            c += dc[dir];
        }

        boolean[] connected = new boolean[4];
        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i];
            int nc = c + dc[i];
            if (nr < 0 || nr >= R || nc < 0 || nc >= C || map[nr][nc] == '.')
                continue;

            char pipe = map[nr][nc];
            if (i == 0 && (pipe == '|' || pipe == '+' || pipe == '1' || pipe == '4')) connected[0] = true;
            else if (i == 1 && (pipe == '|' || pipe == '+' || pipe == '2' || pipe == '3')) connected[1] = true;
            else if (i == 2 && (pipe == '-' || pipe == '+' || pipe == '1' || pipe == '2')) connected[2] = true;
            else if (i == 3 && (pipe == '-' || pipe == '+' || pipe == '3' || pipe == '4')) connected[3] = true;
        }

        System.out.print((r + 1) + " " + (c + 1) + " ");
        if (connected[0] && connected[1] && connected[2] && connected[3]) System.out.println("+");
        else if (connected[0] && connected[1]) System.out.println("|");
        else if (connected[2] && connected[3]) System.out.println("-");
        else if (connected[1] && connected[3]) System.out.println("1");
        else if (connected[0] && connected[3]) System.out.println("2");
        else if (connected[0] && connected[2]) System.out.println("3");
        else if (connected[1] && connected[2]) System.out.println("4");
    }
}

File: src/baekjoon/dp/b2931/b2931.py
import sys

def solve():
    R, C = map(int, sys.stdin.readline().split())
    grid = [list(sys.stdin.readline().strip()) for _ in range(R)]

    # 출발지 찾기
    start_r, start_c = -1, -1
    for r in range(R):
        for c in range(C):
            if grid[r][c] == 'M':
                start_r, start_c = r, c
                break
        if start_r != -1:
            break

    # 방향: 0:상, 1:하, 2:좌, 3:우
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]

    # 1. 첫 흐름 찾기
    r, c, direction = -1, -1, -1
    for i in range(4):
        nr, nc = start_r + dr[i], start_c + dc[i]
        if 0 <= nr < R and 0 <= nc < C and grid[nr][nc] not in ['.', 'Z']:
            r, c, direction = nr, nc, i
            break

    # 2. 시물레이션
    while grid[r][c] != '.':
        block = grid[r][c]

        if block == '1': # └ (아래, 오른쪽)
            # 진입: 위(dir=0) 또는 왼쪽(dir=2)
            if direction == 0: direction = 3 # 위->오른쪽
            else: direction = 1 # 오른쪽->아래
        elif block == '2': # ┌ (위, 오른쪽)
            # 진입: 아래(dir=1) 또는 왼쪽(dir=2)
            if direction == 1: direction = 3 # 아래->오른쪽
            else: direction = 0 # 오른쪽->위
        elif block == '3': # ┐ (위, 왼쪽)
            # 진입: 아래(dir=1) 또는 오른쪽(dir=3)
            if direction == 1: direction = 2 # 아래->왼쪽
            else: direction = 0 # 왼쪽->위
        elif block == '4': # ┘ (아래, 왼쪽)
            # 진입: 위(dir=0) 또는 오른쪽(dir=3)
            if direction == 0: direction = 2 # 위->왼쪽
            else: direction = 1 # 왼쪽->아래

        # '|', '-', '+'는 방향이 바뀌지 않음

        r += dr[direction]
        c += dc[direction]

    # 3.누락된 파이프 결정
    # r, c가 누락된 위치
    connected = [False] * 4 # 상, 하, 좌, 우

    for i in range(4):
        nr, nc = r + dr[i], c + dc[i]
        if not (0 <= nr < R and 0 <= nc < C and grid[nr][nc] != '.'):
            continue

        pipe = grid[nr][nc]
        # i 방향에 있는 파이프가 (r, c) 쪽으로 연결되는지 확인, i 방향으로 흐르고 있다고 생각하면 됨
        if i == 0 and pipe in ['|', '+', '1', '4']: connected[0] = True # 위에서 옴 -> 온다는 것 보다는 연결로 이해하는게 나을듯
        elif i == 1 and pipe in ['|', '+', '2', '3']: connected[1] = True # 아래서 옴
        elif i == 2 and pipe in ['-', '+', '1', '2']: connected[2] = True # 좌측서 옴
        elif i == 3 and pipe in ['-', '+', '3', '4']: connected[3] = True # 우측서 옴

    print(r + 1, c + 1, end=' ')

    if connected == [True, True, True, True]: print('+')
    elif connected == [True, True, False, False]: print('|')
    elif connected == [False, False, True, True]: print('-')
    elif connected == [False, True, False, True]: print('1')
    elif connected == [True, False, False, True]: print('2')
    elif connected == [True, False, True, False]: print('3')
    elif connected == [False, True, True, False]: print('4')

solve()